File: apps/desktop/services/cache.go
package services

import (
	"database/sql"
	"encoding/json"
	"fmt"
)

// CacheService 提供基于 SQLite 的缓存操作（最近目录、收藏夹、设置、标签颜色）
type CacheService struct {
	db *sql.DB
}

// NewCacheService 创建 CacheService 实例
func NewCacheService(db *sql.DB) *CacheService {
	return &CacheService{db: db}
}

// SettingsInfo 内部使用的设置（与 main 包的 Settings 对应）
type SettingsInfo struct {
	Theme    string
	Language string
}

// --- 最近打开的目录 ---

// GetRecentFolders 获取最近打开的目录列表（按打开时间降序，最多 20 条）
func (s *CacheService) GetRecentFolders() ([]string, error) {
	rows, err := s.db.Query(`
		SELECT path FROM recent_folders
		ORDER BY opened_at DESC
		LIMIT 20
	`)
	if err != nil {
		return nil, fmt.Errorf("查询最近目录失败: %w", err)
	}
	defer rows.Close()

	var paths []string
	for rows.Next() {
		var path string
		if err := rows.Scan(&path); err != nil {
			return nil, err
		}
		paths = append(paths, path)
	}
	if paths == nil {
		paths = []string{}
	}
	return paths, rows.Err()
}

// AddRecentFolder 添加或更新最近打开的目录（upsert 更新 opened_at）
func (s *CacheService) AddRecentFolder(path string) error {
	_, err := s.db.Exec(`
		INSERT INTO recent_folders (path, opened_at)
		VALUES (?, datetime('now'))
		ON CONFLICT(path) DO UPDATE SET opened_at = datetime('now')
	`, path)
	if err != nil {
		return fmt.Errorf("添加最近目录失败: %w", err)
	}

	// 保留最近 50 条记录，删除旧的
	_, err = s.db.Exec(`
		DELETE FROM recent_folders
		WHERE path NOT IN (
			SELECT path FROM recent_folders
			ORDER BY opened_at DESC
			LIMIT 50
		)
	`)
	return err
}

// --- 收藏夹 ---

// GetFavorites 获取所有已收藏的路径列表
func (s *CacheService) GetFavorites() ([]string, error) {
	rows, err := s.db.Query(`SELECT path FROM favorites ORDER BY path`)
	if err != nil {
		return nil, fmt.Errorf("查询收藏夹失败: %w", err)
	}
	defer rows.Close()

	var paths []string
	for rows.Next() {
		var path string
		if err := rows.Scan(&path); err != nil {
			return nil, err
		}
		paths = append(paths, path)
	}
	if paths == nil {
		paths = []string{}
	}
	return paths, rows.Err()
}

// ToggleFavorite 切换路径的收藏状态：已收藏则取消，未收藏则添加
func (s *CacheService) ToggleFavorite(path string) error {
	// 先检查是否已收藏
	var count int
	if err := s.db.QueryRow(`SELECT COUNT(*) FROM favorites WHERE path = ?`, path).Scan(&count); err != nil {
		return fmt.Errorf("查询收藏状态失败: %w", err)
	}

	if count > 0 {
		_, err := s.db.Exec(`DELETE FROM favorites WHERE path = ?`, path)
		if err != nil {
			return fmt.Errorf("取消收藏失败: %w", err)
		}
	} else {
		_, err := s.db.Exec(`INSERT INTO favorites (path) VALUES (?)`, path)
		if err != nil {
			return fmt.Errorf("添加收藏失败: %w", err)
		}
	}
	return nil
}

// --- 设置 ---

// GetSettings 读取用户设置，若不存在则返回默认值
func (s *CacheService) GetSettings() (*SettingsInfo, error) {
	settings := &SettingsInfo{
		Theme:    "system",
		Language: "zh",
	}

	rows, err := s.db.Query(`SELECT key, value FROM settings`)
	if err != nil {
		return nil, fmt.Errorf("查询设置失败: %w", err)
	}
	defer rows.Close()

	for rows.Next() {
		var key, value string
		if err := rows.Scan(&key, &value); err != nil {
			return nil, err
		}
		switch key {
		case "theme":
			settings.Theme = value
		case "language":
			settings.Language = value
		}
	}
	return settings, rows.Err()
}

// SaveSettings 保存用户设置（逐字段 upsert）
func (s *CacheService) SaveSettings(settings *SettingsInfo) error {
	kvPairs := map[string]string{
		"theme":    settings.Theme,
		"language": settings.Language,
	}

	tx, err := s.db.Begin()
	if err != nil {
		return fmt.Errorf("开启事务失败: %w", err)
	}
	defer tx.Rollback() //nolint:errcheck

	for k, v := range kvPairs {
		_, err := tx.Exec(`
			INSERT INTO settings (key, value)
			VALUES (?, ?)
			ON CONFLICT(key) DO UPDATE SET value = excluded.value
		`, k, v)
		if err != nil {
			return fmt.Errorf("保存设置 %s 失败: %w", k, err)
		}
	}

	return tx.Commit()
}

// --- 标签颜色 ---

// GetTagColors 获取所有标签颜色映射（tag -> color）
func (s *CacheService) GetTagColors() (map[string]string, error) {
	rows, err := s.db.Query(`SELECT tag, color FROM tag_colors`)
	if err != nil {
		return nil, fmt.Errorf("查询标签颜色失败: %w", err)
	}
	defer rows.Close()

	colors := make(map[string]string)
	for rows.Next() {
		var tag, color string
		if err := rows.Scan(&tag, &color); err != nil {
			return nil, err
		}
		colors[tag] = color
	}
	return colors, rows.Err()
}

// SetTagColor 设置指定标签的颜色（upsert）
func (s *CacheService) SetTagColor(tag, color string) error {
	_, err := s.db.Exec(`
		INSERT INTO tag_colors (tag, color)
		VALUES (?, ?)
		ON CONFLICT(tag) DO UPDATE SET color = excluded.color
	`, tag, color)
	if err != nil {
		return fmt.Errorf("设置标签颜色失败: %w", err)
	}
	return nil
}

// jsonValue 辅助函数：将任意值序列化为 JSON 字符串（用于扩展存储）
func jsonValue(v any) (string, error) {
	b, err := json.Marshal(v)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

File: apps/desktop/src-tauri/src/commands/cache.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use tauri::State;

use crate::AppState;

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum Theme {
    Light,
    Dark,
    System,
}

impl Theme {
    fn as_str(&self) -> &'static str {
        match self {
            Self::Light => "light",
            Self::Dark => "dark",
            Self::System => "system",
        }
    }

    fn from_str(s: &str) -> Self {
        match s {
            "light" => Self::Light,
            "dark" => Self::Dark,
            _ => Self::System,
        }
    }
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum Language {
    En,
    Zh,
}

impl Language {
    fn as_str(&self) -> &'static str {
        match self {
            Self::En => "en",
            Self::Zh => "zh",
        }
    }

    fn from_str(s: &str) -> Self {
        match s {
            "en" => Self::En,
            _ => Self::Zh,
        }
    }
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Settings {
    pub theme: Theme,
    pub language: Language,
}

impl Default for Settings {
    fn default() -> Self {
        Self {
            theme: Theme::System,
            language: Language::Zh,
        }
    }
}

/// 通用查询辅助：执行返回单列字符串列表的 SQL
fn query_string_list(conn: &rusqlite::Connection, sql: &str) -> Result<Vec<String>, String> {
    let mut stmt = conn.prepare(sql).map_err(|e| e.to_string())?;
    let rows = stmt
        .query_map([], |row| row.get::<_, String>(0))
        .map_err(|e| e.to_string())?;
    rows.map(|r| r.map_err(|e| e.to_string())).collect()
}

// ---- 最近文件夹 ----

#[tauri::command]
pub fn get_recent_folders(state: State<AppState>) -> Result<Vec<String>, String> {
    let conn = state.db.lock();
    query_string_list(&conn, "SELECT path FROM recent_folders ORDER BY accessed_at DESC LIMIT 20")
}

#[tauri::command]
pub fn add_recent_folder(state: State<AppState>, path: String) -> Result<(), String> {
    let conn = state.db.lock();
    conn.execute(
        "INSERT INTO recent_folders (path, accessed_at) VALUES (?1, datetime('now'))
         ON CONFLICT(path) DO UPDATE SET accessed_at = datetime('now')",
        [&path],
    )
    .map_err(|e| e.to_string())?;
    // 只保留最近 20 条
    conn.execute(
        "DELETE FROM recent_folders WHERE path NOT IN (
             SELECT path FROM recent_folders ORDER BY accessed_at DESC LIMIT 20
         )",
        [],
    )
    .map_err(|e| e.to_string())?;
    Ok(())
}

// ---- 设置 ----

#[tauri::command]
pub fn get_settings(state: State<AppState>) -> Result<Settings, String> {
    let conn = state.db.lock();
    let theme = get_setting(&conn, "theme")
        .map(|s| Theme::from_str(&s))
        .unwrap_or(Theme::System);
    let language = get_setting(&conn, "language")
        .map(|s| Language::from_str(&s))
        .unwrap_or(Language::Zh);
    Ok(Settings { theme, language })
}

#[tauri::command]
pub fn save_settings(state: State<AppState>, settings: Settings) -> Result<(), String> {
    let conn = state.db.lock();
    set_setting(&conn, "theme", settings.theme.as_str()).map_err(|e| e.to_string())?;
    set_setting(&conn, "language", settings.language.as_str()).map_err(|e| e.to_string())?;
    Ok(())
}

fn get_setting(conn: &rusqlite::Connection, key: &str) -> Option<String> {
    conn.query_row(
        "SELECT value FROM settings WHERE key = ?1",
        [key],
        |row| row.get(0),
    )
    .ok()
}

fn set_setting(conn: &rusqlite::Connection, key: &str, value: &str) -> rusqlite::Result<()> {
    conn.execute(
        "INSERT INTO settings (key, value) VALUES (?1, ?2)
         ON CONFLICT(key) DO UPDATE SET value = ?2",
        [key, value],
    )?;
    Ok(())
}

// ---- 标签颜色 ----

#[tauri::command]
pub fn get_tag_colors(state: State<AppState>) -> Result<HashMap<String, String>, String> {
    let conn = state.db.lock();
    let mut stmt = conn
        .prepare("SELECT tag, color FROM tag_colors")
        .map_err(|e| e.to_string())?;
    let rows = stmt
        .query_map([], |row| Ok((row.get::<_, String>(0)?, row.get::<_, String>(1)?)))
        .map_err(|e| e.to_string())?;
    let mut map = HashMap::new();
    for r in rows {
        let (tag, color) = r.map_err(|e| e.to_string())?;
        map.insert(tag, color);
    }
    Ok(map)
}

/// 合法的 hex 颜色格式：#RRGGBB（与前端 TAG_COLOR_REGEX 保持一致）
fn is_valid_hex_color(s: &str) -> bool {
    s.len() == 7
        && s.starts_with('#')
        && s[1..].bytes().all(|b| b.is_ascii_hexdigit())
}

#[tauri::command]
pub fn set_tag_color(state: State<AppState>, tag: String, color: String) -> Result<(), String> {
    if !is_valid_hex_color(&color) {
        return Err("颜色格式不合法，需要 #RRGGBB 格式".to_string());
    }
    let conn = state.db.lock();
    conn.execute(
        "INSERT INTO tag_colors (tag, color) VALUES (?1, ?2)
         ON CONFLICT(tag) DO UPDATE SET color = ?2",
        [&tag, &color],
    )
    .map_err(|e| e.to_string())?;
    Ok(())
}

// ---- 收藏夹 ----

#[tauri::command]
pub fn get_favorites(state: State<AppState>) -> Result<Vec<String>, String> {
    let conn = state.db.lock();
    query_string_list(&conn, "SELECT path FROM favorites ORDER BY created_at DESC")
}

#[tauri::command]
pub fn toggle_favorite(state: State<AppState>, path: String) -> Result<(), String> {
    let conn = state.db.lock();
    let exists: bool = conn
        .query_row(
            "SELECT COUNT(*) FROM favorites WHERE path = ?1",
            [&path],
            |row| row.get::<_, i64>(0),
        )
        .map(|c| c > 0)
        .map_err(|e| e.to_string())?;

    if exists {
        conn.execute("DELETE FROM favorites WHERE path = ?1", [&path])
            .map_err(|e| e.to_string())?;
    } else {
        conn.execute(
            "INSERT INTO favorites (path) VALUES (?1)",
            [&path],
        )
        .map_err(|e| e.to_string())?;
    }
    Ok(())
}
